# TO DO!!!

-Boolean operations

-part 1 of python basics

# Computer Programming for Musicians 1

Hello! Welcome to Computer Programming for Musicians! 

TBD

# Who we are

Hi Liam

# What you will learn!

TBD

That being said, you should also establish personal goals here! 

# Todays Topics

Today will cover a lot of introductory material! We will learn about **Python**, the language we will use for this short course. We will write our first program, and learn the ins and outs of what sort of things we can create. More specifically

0) Intro ish stuff

1) Python Basics

2) Data Types

3) Operators

These will all be well labelled and numbers for your convenience!

# Anyone can code

As Gusteau says in Ratatoullie, "Anyone can cook." And coding is kind of like cooking so therefore everyone can code! Don't be afraid of any and all big words that pop up. As musicians, you see words like "sforzando, coda, New York Philharmonic." These big words are extremely unfamiliar to the non-musician, but you know them to basically be labels and words to describe common things. We say **marcato** for "marked or distinguished" the same way a programmer says **algorithm** for a set of steps and operations. 

These key words will be **bolded** as we go through. Don't worry too much about the terminology - focus more on the concepts.

And have fun! Programming can be fun!

# 0) Intro ish stuff

This section will discuss what Python is and what this Jupyter Notebook thing is.

## 0.1) What is Python?

You've probably heard about different **languages** in computer programming. There's **Python**, Julia, C, C++, Java, and many more. There are different reasons to use different ones, but for this short course, we will use **Python**. There's a few reasons for this, but the main reason is that Python is a lot more intuitive than some of these other languages. If you're curious you can look into the specs of the language (it's actually kind of slow). But nowadays, many companies use it! 

Python is a great stepping stone to learn how to code! If you end up pursuing the route of being a programmer, you'll eventually learn other languages. However, we firmly believe that having strong grounding in Python is a great way to learn to *learn* how to learn other languages!

## 0.2) What is this Jupyter Notebook?

You've probably seen these hacker scenes from movies that look like this:

### INSERT TONY STARK COMPUTER THING HERE 

In reality, it's probably not *this cool* all the time, but it is still cool! Oftentimes we write code in a python (.py) file and then run it in terminal or command prompt.

Ok...that's a bit confusing. We'll eventually show you how this is done, but in the meantime, let's just write code! This thing you're looking at is called a **Jupyter Notebook**. It let's us write text and also write code blocks!

Check this out, click on the box below and press [Shift + Enter].

In [2]:
# Click here and press shift + enter!
print("If you did it properly, right below it should print out: 'If you did it properly, right below it should print out' ")

If you did it properly, right below it should print out: 'If you did it properly, right below it should print out' 


It's a super convenient way to write and try code! So, thus, we will use this for this short course!

# 1) Python Basics

So, we should probably cover some basics! Within the next few minutes, we will actually write our first program! So how do we do that?

#### NEED TO FINISH

In [ ]:
# Printing
# etc/
# ==, !=, if, else.

# Shallow and deep copies?
# Like A =10, B = 20. A < B. change A = B, can change

# Operators (+, -, *, /, %, >, <, ==, !=)

# input - so people can type


# 2) Data Types

We will now cover something called **Data Types**, which is basically a fancy word to describe what something is. Why do we care about this?

Say you're about to leave your apartment to take a walk, and you want to know what the weather is like outside. How can you do so?

"Is it hot ouside?" "Yes!"

"What's the weather like?" "It's pretty good! Warm!"

"What's the temperature?" "15 degrees (celsius of course)"

"What's the temperature to a lot of decimals? "15.155345 degrees."

So, it looks like the type of question we ask affects the type of response! This is, in essence, what data types are!



## 2.1) The specific data types: Intro

There are a few data types in Python, but we will focus on four: **Booleans, Strings, Integers, Floats**

We can check the type of something by using *type()*! Take a look at the code block below. We create four examples for you to see what is printed out. When we print the type of the variable, we obtain 'bool', 'str', 'int', and 'float.'

In [1]:
# You'll learn these things later, but just to give you a quick taste...

Boolean_example = True
String_example = "word"
Integer_example = 10
Float_example = 10.25

print(type(Boolean_example))
print(type(String_example))
print(type(Integer_example))
print(type(Float_example))

<class 'bool'>
<class 'str'>
<class 'int'>
<class 'float'>




## 2.2) Booleans

A **Boolean** can be thought of as a Yes/No or True/False. To make a Boolean type, we will take a variable and assign it as True or False. Sometimes, we even use 0/1 (with 0 being False and 1 being True). So, in our example above, a boolean is the following:

"Is it hot outside?"

"True!"

This is a boolean!

In the example below, we create four booleans!

In [9]:
# Boolean Examples!


Bool1 = True
Bool2 = False

Bool3 = 1
Bool4 = 0

# Let's verify that this works!

# First we check that Bool1 is True.
print("Is Bool1 True?")
print(Bool1 == True)

print("Is Bool2 True?")
print(Bool2 == True)

print("Is Bool2 False?")
print(Bool2 == False)

print("Is Bool2 0?")
print(Bool2 == 0)

# Let's check out this 0 and 1 thing now!

print("Is Bool3 True?")
print(Bool3 == False)

print("Is Bool4 True?")
print(Bool4 == False)

print("Is Bool4 False?")
print(Bool4 == False)


Is Bool1 True?
True
Is Bool2 True?
False
Is Bool2 False?
True
Is Bool2 0?
True
Is Bool3 True?
False
Is Bool4 True?
True
Is Bool4 False?
True


**WARNING**: *make sure you capitalize your True and False! Or else, it will come in as an error.*

Also, what you'll notice is that when we do a Bool4 == False, it prints out True or False! So the result of this comparison itself is a Boolean!

### Boolean Practice!

Let's work with some Booleans!

In [ ]:
"""
Below are two variables that haven't been defined yet!
Try the following!

-For my_first_boolean, make it a True or False value!
-For my_second_boolean, give it a value using only 0 or 1.
-For I_practiced_today, be honest and write whether if it is true or not!

Then, like we did in the example above, write some code to check and print out if these booleans are 
"""

# Variables
my_first_boolean = 
my_second_boolean = 
I_practiced_today = 

# Check if my_first_boolean is True! Check if it is False!


# Check if my_second_boolean is 0. Check if it's 1! Do this without using 0 or 1! (Hint: Remember that True and 1 are equivalent)


# Check if you practiced today! Uh oh!



In [ ]:
## BOOLEAN OPERTIONS: True + True and whatever

## 2.3) Strings

A **string** can be thought of as a "string of letters, characters, or words." In the weather example, this could be the response: "It's pretty good! Warm!"

So how do we make a string? Conveniently, we actually just need to put things in quotes! Check this out!


In [11]:
# An example of a string!

String1 = "This is a string! It can be very long or as short as"
String2 = "a"
String3 = "single character!"
String4 = "The phrasing was okay, but try less bow. The second note was out of tune."

print(String1)
print(String2)
print(String3)
print(String4)

This is a string! It can be very long or as short as
a
single character!
The phrasing was okay, but try less bow. The second note was out of tune.


Notice that we can print these out immediately!

### String Operations

Let's cover some of the things we can do with strings! 

Firstly, we can add strings! We can take two (or many more) strings and add them together to create a longer string!


In [15]:
# Adding example

first_word = "I"
second_word = "am"
third_word = "out"
fourth_word = "of"
fifth_word = "tune"
period = "."
space = " "

sentence = first_word + space + second_word + space + third_word + space + fourth_word + space + fifth_word + period
print(sentence)


I am out of tune.


We don't necessarily need to define variables to add strings either! We can also mix and match!

In [19]:
# An example of mixing and matching!

mystery_word = " Cool!"
sentence2 = "This is an " + "example of adding" + " without defining variables!" + mystery_word

print(sentence2)

This is an example of adding without defining variables! Cool!


Another thing we can do is pick our **characters** from a string. We do this by **indexing**: we can take the variable and write [some number]. For example:

In [21]:
# indexing strings

string1 = "This is a super long string and I will use [0] to rip off 'T', and [5] to rip off 'i' space!"

print(string1[0])
print(string1[5])

# We can also define an integer (to be defined later) to rip off characters!

number = 10
print(string1[number])

T
i
s


*Notice that the first letter is 0, NOT 1! Python is zero indexed!* We can also define a **range** to rip of part of a string by using [start: end]! For example:

In [25]:
# Taking a chunk of a string

string2 = "I apple dog. I'll bring out apple!"
print(string2[2: 8]) # notice here that apple takes the 2nd to 7th characters. However, this string ends at 8!

apple 


As mentioned in the comment above, the 'end' is really the last spot plus one!

We can also find the length of a string by using the len() function. For example:

In [26]:
# The result of len() will be an integer
length = len(string2)
print(length)

34


### String Practice!

And now, it's your turn to get familiar with this! Here, we will create, add, and manipulate strings for mastery!

In [ ]:
"""
Below is a variable called jury_comment. Try the following!

-Write out a comment you got from your jury (good or bad)!
-Rip off the third word from the comment
-Find the lengths of the jury comment and the third word
-Verify that the jury_comment is longer. (Hint: use > or <. It should return a boolean)
-Rip off the third letter of the third word. If the word is shorter than 3 letters, just pick your favorite letter!
-Add the third word to the end of the jury comment. (If you want it to be pretty, you may need to add a space!)


"""

# write your jury comment
jury_comment = 


# using the jury_comment above, fill in the thing in the brackets to get the third word.
third_word = jury_comment[ : ]


# find the length of the jury_comment and the third_word
jury_comment_length = 
third_word_length =
is_comment_longer = 


# from the third word, take the third letter! (If the word is shorter than 3 letters, take your favorite letter)
third_letter_of_third_word = 


# Add the third word to the end of the jury comment. 
jury_comment_and_third_word = 

# These print statements here are to verify that your results are correct!
print(jury_comment)
print(third_word)
print(jury_comment_length)
print(third_word_length)
print(is_comment_longer)
print(third_letter_of_third_word)
print(jury_comment_and_third_word)

## 2.4) Integers and Floats

**Integers** are the numbers we use most often in every day life (I think). They are the whole numbers we count with: 0, 1, 2, and so on. We can have both positive and negative integers. 

In Python, we can denote an integer quite easily!  We just type the number!


In [2]:
int1 = 1
int2 = 2
int3 = -2

print(type(int3)) # just to check and make sure we get an integer

<class 'int'>


**Floats** are also numbers, but unlike integers, they are not always *whole* numbers. Think of a float as "any number." For example, pi, 3.1415926..., can be expressed as a float. To create a float, we just need to include a decimal point!

In [1]:
float1 = 3.1415
float2 = 0.33
float3 = 1.

print(type(float3))

<class 'float'>


You'll notice that the last float, 1., is a float, and not an integer! Including the decimal point at the end makes it a float type and not an integer. (We could also write 1.0, 1.00, etc.)

While 1 and 1. are, to us, practically the same, on a computer we want to make sure we specify which we are working with! For example, if we are just counting up, integers are fine! (1, 2, 3, ...) 

But maybe we want to calculate how much something costs, and we need those extra decimal points! Now, you might ask "so why don't I just always use a float?" Generally, floats are more espensive - our computers will need to use more resources to keep track of it.

### Integer and Float operations

Let's start by discussing the operations that both integers and floats can use, and then go into float and integer specific ones.

We can add (+), subtract (-), multiply (\*) and divide (/). Let's see some of these in action.


In [5]:
print(1 + 5)
print(1 + 10.)

print(4 - 3)

print(100 * 32)

print(16/5)
print(1/3)

6
11.0
1
3200
3.2
0.3333333333333333


Notice that when we add an integer and a float together, our result is automatically a float! (1 + 10. = 11.0) Also, notice that when we divide two integers, if the result is not a whole number, we get a float (16/5 = 3.2).

Also, when we do 1/3, we get 0.3333333333333333. In reality, this should repeat forever, but our computer doesn't really have infinite space.

## 2.6) Type Casting

## 2.X) Examples and Programming Practice

Try these out! There could be multiple right answers?

**What data type should we use for bank accounts?**

**What data type should we use to indicate what place you are in a line at a supermarket?**

**What data type should we use for a name tag?**

**What data type should we use for a coin flip?**